### PR TITLE
Initial creation of the local serializer.

### PR DESCRIPTION
**Pull request by rsgowman**
Added a single, basic test as a motivator.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/local/serializer.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+/**
+ * @brief Serializer for values stored in the LocalStore
+ *
+ * Note tha local::Serializer currently delegates to the remote::Serializer (for
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/serializer.h]: Nit: s/tha/that.

File: Firestore/core/src/firebase/firestore/local/serializer.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+#include "Firestore/core/src/firebase/firestore/nanopb/writer.h"
+#include "Firestore/core/src/firebase/firestore/remote/serializer.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+/**
+ * @brief Serializer for values stored in the LocalStore
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/serializer.h]: Ultranit: add a full stop.

File: Firestore/core/src/firebase/firestore/local/serializer.h [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * Note tha local::Serializer currently delegates to the remote::Serializer (for
+ * the Firestore v1beta1 RPC protocol) to save implementation time and code
+ * duplication. We'll need to revisit this when the RPC protocol we use diverges
+ * from local storage.
+ */
+class Serializer {
+ public:
+  Serializer(const remote::Serializer& rpc_serializer)
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/serializer.h]: Explicit? If this is deliberate, please add a brief comment stating that.

File: Firestore/core/src/firebase/firestore/local/serializer.cc [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/local/serializer.h"
+#include "Firestore/Protos/nanopb/firestore/local/maybe_document.nanopb.h"
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/serializer.cc]: Nit: I think the style guide wants a blank line after including "own" header.

File: Firestore/core/src/firebase/firestore/local/serializer.cc [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        const model::Document& doc =
+            static_cast<const model::Document&>(maybe_doc);
+        rpc_serializer_.EncodeDocument(writer, doc.key(),
+                                       doc.data().object_value());
+      });
+      return;
+
+    case model::MaybeDocument::Type::NoDocument:
+      // TODO(rsgowman)
+      abort();
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/serializer.cc]: Optional: `#include <cstdlib>`?

File: Firestore/core/src/firebase/firestore/local/serializer.cc [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+void Serializer::EncodeMaybeDocument(
+    Writer* writer, const model::MaybeDocument& maybe_doc) const {
+  switch (maybe_doc.type()) {
+    case model::MaybeDocument::Type::Document:
+      writer->WriteTag(
+          {PB_WT_STRING, firestore_client_MaybeDocument_document_tag});
+      writer->WriteNestedMessage([&](Writer* writer) {
+        const model::Document& doc =
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/serializer.cc]: Optional: I think that when casting, it's better to use `auto`:
* type is clear from the statement;
* avoids duplication;
* eliminates potential for double-conversion:
```cpp
double x{};
int y = static_cast<float>(x);
```
Feel free to leave as is, though.

File: Firestore/core/test/firebase/firestore/local/serializer_test.cc [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+using firebase::firestore::model::SnapshotVersion;
+using firebase::firestore::testutil::Doc;
+using firebase::firestore::util::Status;
+using firebase::firestore::util::StatusOr;
+using google::protobuf::util::MessageDifferencer;
+
+// TODO(rsgowman): This is copied from remote/serializer_tests.cc. Refactor.
+#define EXPECT_OK(status) EXPECT_TRUE(StatusOk(status))
+
+// Fixutre for running the local serializer tests.
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/local/serializer_test.cc]: Nit: s/Fixutre/Fixture. Also, I think you can remove the comment altogether, it seems pretty clear what the class does from its base and its code location (optional).

File: Firestore/core/test/firebase/firestore/local/serializer_test.cc [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    EXPECT_TRUE(ok);
+    EXPECT_TRUE(msg_diff.Compare(proto, actual_proto)) << message_differences;
+  }
+
+  void ExpectDeserializationRoundTrip(
+      const MaybeDocument& model,
+      const firestore::client::MaybeDocument& proto,
+      MaybeDocument::Type type) {
+    size_t size = proto.ByteSizeLong();
+    std::vector<uint8_t> bytes(size);
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/local/serializer_test.cc]: Optional: similarly to another PR, I think you can do away with storing `size` as a local variable:
```cpp
std::vector<uint8_t> bytes(proto.ByteSizeLong());
bool status = proto.SerializeToArray(bytes.data(), bytes.size());
```

File: Firestore/core/test/firebase/firestore/local/serializer_test.cc [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    bool status = proto.SerializeToArray(bytes.data(), size);
+    EXPECT_TRUE(status);
+    StatusOr<std::unique_ptr<MaybeDocument>> actual_model_status =
+        serializer.DecodeMaybeDocument(bytes);
+    EXPECT_OK(actual_model_status);
+    std::unique_ptr<MaybeDocument> actual_model =
+        std::move(actual_model_status).ValueOrDie();
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/local/serializer_test.cc]: Oh, `StatusOr::ValueOrDie()` uses reference-qualified overloads, nice.

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 5 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: Done.

File: Firestore/core/src/firebase/firestore/local/serializer.h (hunk 2 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.h]: Done.

File: Firestore/core/src/firebase/firestore/local/serializer.h (hunk 1 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.h]: Done.

File: Firestore/core/src/firebase/firestore/local/serializer.h (hunk 3 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.h]: Done.

File: Firestore/core/test/firebase/firestore/local/serializer_test.cc (hunk 7 above)
Comment (rsgowman) [reply Firestore/core/test/firebase/firestore/local/serializer_test.cc]: Yeah, agreed that it's useless. Removed.

File: Firestore/core/test/firebase/firestore/local/serializer_test.cc (hunk 8 above)
Comment (rsgowman) [reply Firestore/core/test/firebase/firestore/local/serializer_test.cc]: Done.

File: Firestore/core/src/firebase/firestore/local/serializer.cc [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        // 'no_document' and 'document' are part of a oneof. The proto docs
+        // claim that if both are set on the wire, the last one wins.
+        document = nullptr;
+
+        // TODO(rsgowman): Parse the no_document field.
+        abort();
+
+      default:
+        // Unknown tag. According to the proto spec, we need to ignore these.
+        reader->SkipField(tag);
```
Comment (zxu123) [inline Firestore/core/src/firebase/firestore/local/serializer.cc]: I feel the logic is tore apart by the two switches. I'd move the logic here above, i.e. in previous default, just do
```
reader->SkipField(tag);
continue;
```
and put unreachable here.

Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: My reasoning behind this was to have the first switch handle the type verification. (e.g. ensure string fields contains strings, int fields contain ints.) And then the second switch can just focus on parsing the contents of the field. I like this separation but am willing to be out-voted. Also note that while it's not entirely obvious here (due to there only being 2 fields) in other messages there are many more. In these cases, the type verification reduces to:

```
switch(field_number) {
  case tag1:
  case tag3:
  case tag5:
    assert(type == string)
    break;
  case 2:
  case 4:
  case 6:
    assert(type == varint)
    break;
  default:
    //nothing!
    break;
}
```

There's typically only the 3 "real" blocks (assuming you count the empty default block). So the type verification switch statement is usually pretty simple, whereas the parsing one is usually more complex.

Although I haven't gotten to it yet, I suspect the first switch statement could be refactored to a single helper method call: `VerifyTypeForMaybeDocumentField(...)` or similar which might address your concerns somewhat? At the very least, it should make it more obvious what's going on.

Regardless of what we decide, the pattern here is used throughout the serializers, so I'll apply the final resolution (if any) everywhere in a followup.

Options:
1) Two switches. (i.e. as currently written)
2) Refactor into `VerifyTypesForMaybeDocumentField`. Two switches still remain, but not in the same method. The parsing itself might be refactored too to make this method even simpler.
3) Merge two switches (verify types + parse in the same case block.)

Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: Or were you referring only to the 'default' case?

Comment (zxu123) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: I'm OK with the other cases's separation. For the `default` case, the logic is simple and might be combined. It is in the same dogma of type
```
func {
  if (foo) {
     bar;
     return;
  }
  // now in below code, we assume !foo.
  ...
}
```

**Comment by rsgowman** [comment]
Oops. This PR is incomplete. (I'm missing the update time.) Hold off on further reviews for a bit.

File: Firestore/core/src/firebase/firestore/local/serializer.cc [hunk 11]
```diff
@@ -44,8 +47,7 @@ void Serializer::EncodeMaybeDocument(
       writer->WriteTag(
           {PB_WT_STRING, firestore_client_MaybeDocument_document_tag});
       writer->WriteNestedMessage([&](Writer* writer) {
-        const model::Document& doc =
-            static_cast<const model::Document&>(maybe_doc);
+        auto doc = static_cast<const model::Document&>(maybe_doc);
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/serializer.cc]: This should be `const auto&`. Plain `auto` will make a copy here.

The rules for `auto` deduction are similar to template argument deduction:
```cpp
template <typename T>
void foo(T t);
template <typename T>
void bar(const T& t);
const std::string& str_ref = some_global_string;
foo(str_ref); // copies -- T deduces to `std::string`, so the argument is `std::string`
auto foo2 = str_ref; // copies -- auto deduces to `std::string`
bar(str_ref); // no copy -- T deduces to `std::string`, so the argument is `const std::string&`
const auto& bar2 = str_ref; // no copy
```
Sorry about that. If you feel like specifying full type makes it easier to avoid this issue, feel free to revert to how it was before.

Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: Nah, this is just me being dumb and not paying enough attention to what I'm doing. I'll keep the auto here (but will add the &). Thanks for picking up on this.

Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: Now deleted altogether.

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 10 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: Basically, an early exit. I'd like to avoid that to keep all the reading together in the second switch. It should also allow some potential future refactoring, i.e.:

```
DecodeFoo() {
  ...
  while (field = GetNextField()) {
    EnsureTagAndTypeCorrect(field.tag, field.type);  // switch #1 hidden in here
    DecodeField(field);  // switch #2 hidden in here
  }
  ...
}
```

(We could still combine the two default cases and still have the above work if the first function returned some sort of early exit indicator.)

**Comment by rsgowman** [comment]
Should be ready for re-review.

File: Firestore/core/src/firebase/firestore/local/serializer.cc [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+#include "Firestore/core/src/firebase/firestore/model/field_value.h"
+#include "Firestore/core/src/firebase/firestore/model/no_document.h"
+#include "Firestore/core/src/firebase/firestore/nanopb/tag.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+using firebase::firestore::model::ObjectValue;
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/local/serializer.cc]: super-mega-ultra-nit: elsewhere we've put the using statements before the namespace opens.

File: Firestore/core/src/firebase/firestore/local/serializer.h [hunk 13]
```diff
@@ -0,0 +1,109 @@
+/*
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/local/serializer.h]: Xcode tends to have problems with projects containing identical filenames.

Also, this class is called a `LocalSerializer` on the other platforms.

Taken together I suggest renaming this and the .cc to `local_serializer` and the class to `LocalSerializer`.

File: Firestore/core/src/firebase/firestore/local/serializer.cc [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+        // TODO(rsgowman): Parse the no_document field.
+        abort();
+
+      default:
+        // Unknown tag. According to the proto spec, we need to ignore these.
+        reader->SkipField(tag);
+    }
+  }
+
+  if (no_document != nullptr) {
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/local/serializer.cc]: Can't this just be `if (no_document)` ?

File: Firestore/core/src/firebase/firestore/local/serializer.cc [hunk 15]
```diff
@@ -140,6 +140,24 @@ std::unique_ptr<model::MaybeDocument> Serializer::DecodeMaybeDocument(
   }
 }
 
+void Serializer::EncodeDocument(Writer* writer, const model::Document& doc) const {
+  // Encode Document.name
+  writer->WriteTag({PB_WT_STRING, google_firestore_v1beta1_Document_name_tag});
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/serializer.cc]: Can this implementation take advantage of the fact that the format is the same as `remote::Serializer::EncodeDocument` with an additional field appended? Something like:
```cpp
rpc_serializer_.EncodeDocument(writer, doc);
writer->WriteTag({PB_WT_STRING, google_firestore_v1beta1_Document_update_time_tag});
writer->WriteNestedMessage([&](Writer* writer) {
  rpc_serializer_.EncodeVersion(writer, doc.version());
});
```

File: Firestore/core/src/firebase/firestore/local/serializer.cc [hunk 16]
```diff
@@ -140,6 +140,24 @@ std::unique_ptr<model::MaybeDocument> Serializer::DecodeMaybeDocument(
...
+  writer->WriteString(rpc_serializer_.EncodeKey(doc.key()));
+
+  // Encode Document.fields (unless it's empty)
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/serializer.cc]: Question: just curious, can the check for emptiness move to `EncodeObjectMap`, or is there a reason to keep it outside?

File: Firestore/core/src/firebase/firestore/local/serializer.cc [hunk 17]
```diff
@@ -140,6 +140,24 @@ std::unique_ptr<model::MaybeDocument> Serializer::DecodeMaybeDocument(
...
+  const ObjectValue& object_value = doc.data().object_value();
+  if (!object_value.internal_value.empty()) {
+    rpc_serializer_.EncodeObjectMap(writer, object_value.internal_value, google_firestore_v1beta1_Document_fields_tag, google_firestore_v1beta1_Document_FieldsEntry_key_tag, google_firestore_v1beta1_Document_FieldsEntry_value_tag);
+  }
+
+  // Encode Document.update_time
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/serializer.cc]: Question: I presume `create_time` is deliberately ignored. If that's the case, perhaps (optionally) add a brief comment on why `update_time` is serialized but `create_time` isn't?

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 15 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: Splitting parsing of a message like that makes me a little nervous, but I think it should work perfectly fine. (Decoding wouldn't though.)

OTOH, it wouldn't be saving much, and these methods are logically distinct. (The local serializer theoretically shouldn't use *anything* from the remote serializer as the on-disk protos and the on-wire protos are considered different. Even though they largely aren't.) So the long term, probably never reached end point, is that none of these methods call anything in the remote serializer.

Left as is for now.

File: Firestore/core/src/firebase/firestore/local/serializer.cc [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+  while (reader->bytes_left()) {
+    Tag tag = reader->ReadTag();
+    if (!reader->status().ok()) return nullptr;
+
+    // Ensure the tag matches the wire type
+    switch (tag.field_number) {
+      case firestore_client_MaybeDocument_document_tag:
+      case firestore_client_MaybeDocument_no_document_tag:
+        if (tag.wire_type != PB_WT_STRING) {
+          reader->set_status(
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/local/serializer.cc]: Is there any reason not to just return immediately after setting status?

I worry that by not returning early we leave ourselves open to the possibility that a future refactoring will break the return that's currently below.

Additionally it makes the error handling (slightly) harder to reason about since the return value and the status generation aren't paired.

These aren't strong arguments, but it really isn't any extra code to bind these more closely.

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 17 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: The honest answer is because the objc implementation deliberately ignores it. I presume it's because we simply don't care about this field. (Or in other words, the on disk proto has "diverged", but not by enough for us to care.) I'll add a comment to that effect; Gil can correct me if I'm wrong.

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 16 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: Hm. I think that could be moved inside without any troubles. I'll address this in a followup (since this should apply within the remote serializer too, and I've got existing followups that rework this code a bit already.)

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 10 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: I was about to comment separately about the two switch thing, so I'll cast my vote for just a single switch statement.

I think this could be made less onerous by adding a method to the reader that allows you to call it like so:

```c++
reader->RequireWireType(PT_WT_STRING, tag);
```

You can then just make this the first line of each case.

I favor this approach because which type applies and the code to handle it are not independent. It would be too easy to change code here to use e.g. an int and then forget to break out the same case above.

You could also have `ReadNestedMessage` take the parent tag as a parameter and have it validate that the parent has the correct wire type (which AIUI will always be PB_WT_STRING for nested messages).

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 15 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: I'm fine either way, thanks for the explanation.

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 12 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: It's an attempt to follow the advice here:
http://go/totw/119#scope-of-the-alias (https://abseil.io/tips/119)

I'd propose fixing the other instances, but the likelihood of this actually biting us seems... remote.

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 18 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: > Is there any reason not to just return immediately after setting status?

Not that I can recall. I've moved it in here (and will do so throughout the rest of the serializers as I touch them).

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 10 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: Alright; I'll author a followup that merges the double-switch pattern wherever it occurs.

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 14 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: Yup.

File: Firestore/core/src/firebase/firestore/local/serializer.h (hunk 13 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/local/serializer.h]: The Redundant Department of Redundancy strikes again.  :(

File: Firestore/core/src/firebase/firestore/remote/serializer.h [hunk 19]
```diff
@@ -159,13 +164,24 @@ class Serializer {
   }
 
  private:
+  friend class local::LocalSerializer;
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/serializer.h]: Friends aren't a concept we have on the other platforms and there's really no reason not to just make the things the LocalSerializer uses public.

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 12 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: That article also argues for fully qualifying to prevent any ambiguity.

Preventing collisions in the global namespace is a worthy goal. Let's clean this up separately.

File: Firestore/core/src/firebase/firestore/local/serializer.cc (hunk 17 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/local/serializer.cc]: TL;DR: omitting creation time is the right call.

As usual, there's history: `create_time` is a read-only property and it's an error to send it to the backend so the remote serializer must omit it. The Objective-C local serializer implementation evolved out of the remote one and preserved this bias, though the bias is unnecessary.

Fixing this has never been a priority because we have no use for creation time and if we ever do need it we're going to need to deal with the on-device cache not having it when we do. I anticipate never needing this value so we can defer this indefinitely.

File: Firestore/core/test/firebase/firestore/local/CMakeLists.txt [hunk 20]
```diff
@@ -16,7 +16,9 @@ cc_test(
   firebase_firestore_local_test
   SOURCES
     leveldb_key_test.cc
+    serializer_test.cc
```
Comment (wilhuff) [inline Firestore/core/test/firebase/firestore/local/CMakeLists.txt]: `local_serializer_test.cc`

File: Firestore/core/src/firebase/firestore/remote/serializer.h (hunk 19 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: It prevents users of the serializer from calling these (implementation detail) methods. The fact that the local serializer happens to use them at all is just a cheat that we can get away with due to the wire proto and disk proto not having sufficiently diverged yet.

These private methods also leak the implementation (nanopb). Mind you, that leak would also likely prevent any caller from accidentally using them, since they'd need to construct a nanopb::Reader/Writer in order to do so.

All-in-all; I'm still slightly in favour of `friend`, even if the other platforms are unable to take advantage of that.

The alternative is to address this via docs: `/** Do not use. Only public for use by the LocalSerializer. */`. I'm less thrilled, but still find this to be perfectly acceptable. (It's not like this is a public api or anything.)

(Another alternative is to refactor this common logic out to a NanopbSerializerHelper class and use that in both serializers. No friends, no needlessly public methods. But this diverges from the other platforms.)

Was the above convincing? If yes, I'll leave as is. If not, I'll switch the (relevant) private methods to public and add the doc string.

File: Firestore/core/test/firebase/firestore/local/CMakeLists.txt (hunk 20 above)
Comment (rsgowman) [reply Firestore/core/test/firebase/firestore/local/CMakeLists.txt]: Oops. Fixed.

File: Firestore/core/src/firebase/firestore/remote/serializer.h (hunk 19 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: What I disagree with is that need to be hidden from anybody. The serializer *itself* is an implementation detail. There's no reason to prevent anyone that needs to serialize a document to nanopb from doing so.

I also don't think we need to go to any lengths to hide the fact that the serializer uses nanopb. The comparable method in the Objective-C and Android-Java implementations returns an object bound to the specific proto library they use.

In short I think just making this public and being done with it is sufficient.
